pkg/sessions/sql/sql_store.go

```diff
@@ -0,0 +1,92 @@
+package sql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	// load the driver
+	_ "github.com/lib/pq"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/sessions/persistence"
+)
+
+// SessionStore is an implementation of the persistence.Store
+// interface that stores sessions in a SQL database
+type SessionStore struct {
+	pool  *sql.DB
+	table string
+}
+
+// NewSQLSessionStore initialises a new instance of the SessionStore and wraps
+// it in a persistence.Manager
+func NewSQLSessionStore(opts *options.SessionOptions, cookieOpts *options.Cookie) (sessions.SessionStore, error) {
+	table := opts.SQL.TablePrefix + "_sessions"
+	pool, err := sql.Open(opts.SQL.Driver, opts.SQL.DSN)
+	if err != nil {
+		return nil, fmt.Errorf("error constructing SQL client: %w", err)
+	}
+	_, err = pool.Exec(fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s (
+            key text primary key,
+            value bytea,
+            expires timestamp)`, table))
+	if err != nil {
+		return nil, fmt.Errorf("error creating table: %w", err)
+	}
+	_, err = pool.Exec(fmt.Sprintf(`BEGIN;
+        CREATE INDEX IF NOT EXISTS exp_idx ON %s (expires);
+        CREATE OR REPLACE FUNCTION f_delete_expired () RETURNS trigger
+            AS 'BEGIN DELETE FROM TG_TABLE_NAME WHERE expires < current_timestamp; RETURN NULL; END' LANGUAGE PLPGSQL;
+        DROP TRIGGER IF EXISTS t_delete_expired ON %s;
+        CREATE TRIGGER t_delete_expired BEFORE INSERT ON %s
+            EXECUTE PROCEDURE f_delete_expired();
+        COMMIT;`, table, table, table))
+	if err != nil {
+		return nil, fmt.Errorf("error creating trigger: %w", err)
+	}
+	ss := &SessionStore{
+		pool:  pool,
+		table: table,
+	}
+	return persistence.NewManager(ss, cookieOpts), nil
+}
+
+// Save takes a sessions.SessionState and stores the information from it
+// to postgres, and adds a new persistence cookie on the HTTP response writer
+func (store *SessionStore) Save(ctx context.Context, key string, value []byte, exp time.Duration) error {
+	_, err := store.pool.ExecContext(ctx, fmt.Sprintf(
+		`INSERT INTO %s VALUES ($1, $2, current_timestamp + $3)
```

Comment:
I agree with this approach -- we have so little data to manage in the database that any ORM is overkill.

I think I would lean toward a specialized client struct (similar to redis) that owns these raw implementation though.  Then we can wrap the client calls in the SQLStore implementation of SessionStore that has Save/Load/Clear

Reply:
Good idea, I actually have a second patch on top of this one which adds end-to-end healthchecks (ie. actually checks if you can read from the datastore), which would be cleaner with this approach.